online_meetings/report.py

```diff
@@ -36,27 +37,50 @@
 DEFAULT_SLEEP_TIME = ENV.get('DEFAULT_SLEEP_TIME', 10)
 
 
-# Functions
-def get_total_page_count(url: str, headers: Dict[str, Union[str, int]] = None, params: Dict[str, Union[str, int]] = None):
-    if params is None:
-        params = {}
-    if headers is None:
-        headers = {}
+def get_request_retry(url: str, headers: Dict[str, Union[str, int]],
+                      params: Dict[str, Union[str, int]]) -> requests.Response:
+
+    response = requests.request("GET", url, headers=headers, params=params)
+
+    # Rate limited, wait a few seconds
+    if response.status_code == requests.codes.too_many_requests:
+        # This is what the header should be
+        retry_after = response.headers.get("Retry-After")
+        sleep_time = DEFAULT_SLEEP_TIME
+        if retry_after and retry_after.isdigit():
+            logger.warning(
+                f"Received status 429, need to wait for {retry_after}")
+            sleep_time = retry_after
+        else:
+            logger.warning(
+                f"No Retry-After header, setting sleep loop for {DEFAULT_SLEEP_TIME} seconds")
+        while response.status_code == requests.codes.too_many_requests:
+            time.sleep(sleep_time)
+            response = requests.request(
+                "GET", url, headers=headers, params=params)
+
+    # If it's not okay at this point, raise an error
+    if response.status_code != requests.codes.ok:
+        logger.warning(f'Received irregular status code: {response.status_code}')
+        response.raise_for_status()
+    return response
 
+
+# Functions
+def get_total_page_count(url: str, headers: Dict[str, Union[str, int]], params: Dict[str, Union[str, int]]):
     # get the total page count
     total_page_count = 0
-    response = requests.request("GET", f"{url}", headers=headers, params=params)
-    status_code = response.status_code
-    if status_code != 200:
-        logger.warning(f'Received irregular status code: {status_code}')
+    try:
+        response = get_request_retry(url, headers, params)
+    except requests.exceptions.HTTPError:
+        logger.exception('Received irregular status code during request')
+        return 0
+    try:
+        results = json.loads(response.text.encode('utf8'))
+        total_page_count = results['page_count']
+    except json.JSONDecodeError:
+        logger.warning('JSONDecodeError encountered')
```

Comment:
Not sure what weird JSON responses happen with this API, but with Canvas, sometimes the response just is invalid. I  think re-trying in those instances could make sense, so maybe at some point these `try/except`s could get moved into `get_request_retry`?

Reply:
Yeah, this `get_request_retry` doesn't catch the Canvas one yet. get_total_page_count is inside of a get_request_retry pattern though which is mostly written to catch the zoom issue of REALLY short throttling. I'd prefer everything with canvas go through the canvasapi package eventually.

Reply:
I'm just going to mark this one as resolved for now, we probably need better API error handling in a central place.